### PR TITLE
2x more memory efficient Graph-based RNN-T

### PR DESCRIPTION
**Pull request by artbataev**
# What does this PR do ?

- reduces memory usage of Graph-Transducer by 2 times: Triton-based implementation of log-softmax operation for RNN-T 
  - idea: store log probabilities only for blank/target labels (needed to compute RNN-T loss), not for all logits
  - if we calculate full logits (Joint tensor), this is the optimal implementation: most of the memory is used by `logits/+gradients` (2x logits)
  - previous implementation required `logits/+gradients`, `log_probs/+gradients` (4x logits)
- adds an option to return RNN-T graphs from `forward` along with the loss value for further usage (e.g., using alignment)

**Example**: bfloat16, 15 sec, 8x subsampling (~FastConformer), 3 words/sec, vocab size 1024, batch size 32

|                         | Logits  | Extra Memory | Time     |
| ----------------------- | ------- | ------------ | -------- |
| Without Triton (before) | 1.04 GB | 3.20 GB      | 0.08 sec |
| **With Triton (here)**  | 1.04 GB | **1.13 GB**  | **0.06 sec** |

3x extra memory reduction, 2x total memory reduction (4.24 GB -> 2.17 GB).

Convergence: stable, see https://wandb.ai/nvidia/Graph_RNNT_Optimized

**Collection**: [ASR]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 1]
```diff
@@ -167,30 +175,28 @@ def get_graphs_batched(
...
             )
             if self.connect_composed:
-                k2.connect(target_fsas_vec)
+                target_fsas_vec = k2.connect(target_fsas_vec)
         return target_fsas_vec
 
-    def get_logits_indices(self, target_fsas_vec: k2.Fsa, logits_shape: torch.Size) -> torch.Tensor:
+    def get_batch_indices(self, target_fsas_vec: k2.Fsa) -> torch.Tensor:
         """
         Get indices of flatten logits for each arc in the lattices.
```
Comment (GNroy) [inline nemo/collections/asr/parts/k2/graph_transducer.py:191]: Is this description relevant?

**Review by GNroy — APPROVED** [review]
Great optimization, LGTM!

File: nemo/collections/asr/parts/k2/graph_transducer.py (hunk 1 above)
Comment (artbataev) [reply nemo/collections/asr/parts/k2/graph_transducer.py]: Fixed, thanks!

**Review by GNroy — APPROVED** [review]

File: nemo/collections/asr/parts/k2/graph_transducer.py [hunk 2]
```diff
@@ -15,11 +15,17 @@
 import abc
 from contextlib import nullcontext
 from typing import ContextManager
+
 import torch
 import torch.nn.functional as F
 
 from nemo.core.classes.loss import Loss
 from nemo.core.utils.k2_guard import k2
+from nemo.core.utils.optional_libs import TRITON_AVAILABLE
```
Comment (titu1994) [inline nemo/collections/asr/parts/k2/graph_transducer.py:24]: Can this be made into a function, similar to Numba, instead of a hardcoded constant var

File: nemo/core/utils/optional_libs.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+def is_lib_available(name: str) -> bool:
+    """
+    Checks if the library/package with `name` is available in the system
+    NB: try/catch with importlib.import_module(name) requires importing the library, which can be slow.
+    So, `find_spec` should be preferred
+    """
+    return importlib.util.find_spec(name) is not None
+
+
+TRITON_AVAILABLE = is_lib_available("triton")
```
Comment (titu1994) [inline nemo/core/utils/optional_libs.py:27]: Can you make this a funcion as well?

**Review by titu1994 — APPROVED** [review]
Very impressive speedups. Minor comments about util function but can be merged without being a blocker

File: nemo/core/utils/optional_libs.py (hunk 3 above)
Comment (artbataev) [reply nemo/core/utils/optional_libs.py:27]: Thank you! Let's discuss this in Slack, I will make a follow-up PR for this stuff
